tawazi/_dag/helpers.py

```diff
@@ -91,6 +80,82 @@ def get_highest_priority_node(
     return xns_dict[highest_priority_node]
 
 
+class BiDict(Dict[K, V]):
+    """A bidirectional dictionary that raises an error if two keys are mapped to the same value.
+
+    >>> b = BiDict({1: 'one', 2: 'two'})
+    >>> b[1]
+    'one'
+    >>> b.inverse['one']
+    1
+    >>> b[3] = 'three'
+    >>> b.inverse['three']
+    3
+    >>> del b[1]
+    >>> assert 1 not in b
+    >>> b[4] = 'one'
+    >>> b.inverse['one']
+    4
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self.inverse: Dict[V, K] = {}
+        for key, value in self.items():
+            self._setitem_inverse(key, value)
+
+    def _setitem_inverse(self, key: K, value: V) -> None:
+        if value in self.inverse:
+            raise ValueError(f"Value {value} is already in the BiDict")
+        self.inverse[value] = key
+
+    def __setitem__(self, key: K, value: V) -> None:
+        if key in self:
+            del self.inverse[self[key]]
+        super().__setitem__(key, value)
+        self._setitem_inverse(key, value)
```

Comment:
please cherry pick the fix for BIDict https://github.com/mindee/tawazi/blob/2899f194733fbcb93b72a4a65a36590f38456c2f/tawazi/_dag/helpers.py#L100